internal/blobstore/multipart.go

```diff
@@ -0,0 +1,224 @@
+// upload id. It returns ErrNotFound if the upload has been
+// deleted or finished.
+func (s *Store) ListParts(uploadId string) ([]Part, error) {
+	return nil, errgo.New("parts listing not implemented yet")
+	// TODO
+	// read multipart metadata
+	// return parts from that, omitting parts that are currently in progress
+}
+
+// partName returns the blob name of the part with the given
+// uploadId and part number.
+func (s *Store) partName(uploadId string, part int) string {
+	return fmt.Sprintf("%s/%d", uploadId, part)
+}
+
+// initializePart creates the initial record for a part.
+func initializePart(uploadc *mgo.Collection, uploadId string, part int, hash string, size int64) error {
+	partElem := fmt.Sprintf("parts.%d", part)
+	err := uploadc.Update(bson.D{
+		{"_id", uploadId},
+		{"$or", []bson.D{{{
+			partElem, bson.D{{"$exists", false}},
+		}}, {{
+			partElem, bson.D{{"$eq", nil}},
+		}}}},
+	},
+		bson.D{{
+			"$set", bson.D{{partElem, uploadPart{
+				Hash: hash,
+				Size: size,
+			}}},
+		}},
+	)
+	if err == nil || err != mgo.ErrNotFound {
+		return nil
+	}
+	return errgo.New("cannot update initial part record - concurrent upload of the same part?")
+}
+
+// FinishParts completes a multipart upload by joining all the given
+// parts into one blob. The resulting blob can be opened by passing
+// uploadId and the returned multipart index to Open.
+//
+// The part numbers used will be from 0 to len(parts)-1.
+//
+// This does not delete the multipart metadata, which should still be
+// deleted explicitly by calling DeleteParts after the index data is
+// stored.
+func (s *Store) FinishParts(uploadId string, parts []Part) (idx *MultipartIndex, hash string, err error) {
+	// TODO read metadata
+	// if parts don't match uploaded parts, return error.
+	// read all parts in sequence to hash them.
+	// return index derived from metadata and calculated hash.
+	return nil, "", errgo.New("parts listing not implemented yet")
+}
+
+// DeleteExpiredParts deletes any multipart entries
+// that have passed their expiry date.
+func (s *Store) DeleteExpiredParts() error {
+	return errgo.New("not implemented yet")
+}
+
+// DeleteParts deletes all the parts associated with the
+// given upload id. It does nothing if FinishParts
+// has already been called for the given upload id.
+func (s *Store) DeleteParts(uploadId string) error {
+	// TODO
+	// read multipart metadata
+	// delete all parts referenced in that
+	// delete multipart metadata
+	return errgo.New("parts listing not implemented yet")
+}
+
+// MultipartIndex holds the index of all the parts of a multipart blob.
+// It should be stored in an external document along with the
+// blob name so that the blob can be downloaded.
+type MultipartIndex struct {
+	Sizes []uint32
+}
+
+// Part represents one part of a multipart blob.
+type Part struct {
+	Hash string
+}
+
+func (s *Store) uploadC() *mgo.Collection {
+	return s.db.C(s.prefix + ".upload")
```

Comment:
are you allowed . in collection names? will this not make it awkward to use in the mongo shell? I don't know it may be fine.